File: a001_two_sum/golang/main.go
package main

import "fmt"

func main() {
	a := []int{0, 1, 2, 3, 4, 5, 6, 7}
	r := twoSum(a, 11)
	fmt.Println(r[0], r[1]) // 5 6
}

// Given an array of integers, return indices of the two numbers
// such that they add up to a specific target.
// You may assume that each input would have exactly one solution,
// and you may not use the same element twice.
func twoSum(a []int, target int) [2]int {
	m := map[int]int{}
	// R: O(n)
	for i := range a {
		n := a[i]
		key := target - n
		if v, ok := m[key]; ok {
			return [2]int{v, i}
		}
		m[n] = i
	}
	panic("No two sum solution")
}

File: a001_two_sum/rs/src/main.rs
fn main() {
    let a = [0, 1, 2, 3, 4, 5, 6, 7];
    let r = two_sum(&a, 11);
    println!("{:?}", r); // [5 6]
}

// Given an array of integers, return indices of the two numbers
// such that they add up to a specific target.
// You may assume that each input would have exactly one solution,
// and you may not use the same element twice.
pub fn two_sum(a: &[isize], target: isize) -> [usize; 2] {
    let mut map = std::collections::HashMap::new();
    // R: O(n)
    for (i, n) in a.iter().enumerate() {
        let key = target - n;
        if map.contains_key(&key) {
            return [map[&key], i];
        }
        map.insert(n, i);
    }
    panic!("No solution");
}
